File: Chap6-GraphicsViewFramework/6-8ResizableItemDemo/handleitem.py
from PySide6.QtWidgets import QGraphicsRectItem
from PySide6.QtCore import QPointF, QRectF
from PySide6.QtGui import QPainter

class HandleItem(QGraphicsRectItem):
    """
    Handle item for resizing items in a graphics scene
    """
    # Enumeration for handle positions
    TopLeft = 0
    TopRight = 1
    BottomRight = 2
    BottomLeft = 3
    
    def __init__(self, position):
        """
        Initialize the handle item with a specified position
        
        Args:
            position: One of the position constants (TopLeft, TopRight, etc.)
        """
        super().__init__()
        self.handle_position = position
        self.setFlag(QGraphicsRectItem.ItemIsMovable)
    
    def mouseMoveEvent(self, event):
        """
        Handle the mouse move event to resize the parent item
        
        Args:
            event: QGraphicsSceneMouseEvent
        """
        if self.handle_position == HandleItem.TopLeft:
            # Try to cast the parent to the right type
            from resizablepixmapitem import ResizablePixmapItem
            rect_item = self.parentItem()
            
            if isinstance(rect_item, ResizablePixmapItem):
                bounding_frame_rect = rect_item.selectorFrameBounds()
                bounding_frame_rect.setTop(event.pos().y())
                bounding_frame_rect.setLeft(event.pos().x())
                rect_item.setSelectorFrameBounds(bounding_frame_rect)
        
        elif self.handle_position == HandleItem.TopRight:
            from resizablerectitem import ResizableRectItem
            rect_item = self.parentItem()
            
            if isinstance(rect_item, ResizableRectItem):
                bounding_frame_rect = rect_item.selectorFrameBounds()
                bounding_frame_rect.setTop(event.pos().y())
                bounding_frame_rect.setRight(event.pos().x())
                rect_item.setSelectorFrameBounds(bounding_frame_rect)
        
        elif self.handle_position == HandleItem.BottomRight:
            from resizablerectitem import ResizableRectItem
            rect_item = self.parentItem()
            
            if isinstance(rect_item, ResizableRectItem):
                bounding_frame_rect = rect_item.selectorFrameBounds()
                bounding_frame_rect.setRight(event.pos().x())
                bounding_frame_rect.setBottom(event.pos().y())
                rect_item.setSelectorFrameBounds(bounding_frame_rect)
        
        elif self.handle_position == HandleItem.BottomLeft:
            from resizablerectitem import ResizableRectItem
            rect_item = self.parentItem()
            
            if isinstance(rect_item, ResizableRectItem):
                bounding_frame_rect = rect_item.selectorFrameBounds()
                bounding_frame_rect.setBottom(event.pos().y())
                bounding_frame_rect.setLeft(event.pos().x())
                rect_item.setSelectorFrameBounds(bounding_frame_rect)

File: 16-GraphicsViewFramework/08_ResizableItemDemo/handleitem.cpp
#include "handleitem.h"
#include "resizablerectitem.h"
#include "resizablepixmapitem.h"
#include "resizablestaritem.h"
#include <QGraphicsSceneMouseEvent>

HandleItem::HandleItem(Position position) :
    handlePosition(position)
{
    setFlag(QGraphicsItem::ItemIsMovable);

}

void HandleItem::mouseMoveEvent(QGraphicsSceneMouseEvent *event)
{
    switch (handlePosition) {

    case TopLeft:
    {
       // ResizableRectItem * rectItem = dynamic_cast<ResizableRectItem *>( parentItem());
      //  ResizablePixmapItem * rectItem = dynamic_cast<ResizablePixmapItem *>( parentItem());
        ResizablePixmapItem * rectItem = dynamic_cast<ResizablePixmapItem *>( parentItem());
        if(rectItem){

            QRectF boundingFrameRect = rectItem->selectorFrameBounds();

            boundingFrameRect.setTop(event->pos().y());
            boundingFrameRect.setLeft(event->pos().x());

            rectItem->setSelectorFrameBounds(boundingFrameRect);
        }

    }
        break;

    case TopRight:
    {
        ResizableRectItem * rectItem = dynamic_cast<ResizableRectItem *>( parentItem());
        if(rectItem){

            QRectF boundingFrameRect = rectItem->selectorFrameBounds();

            boundingFrameRect.setTop(event->pos().y());
            boundingFrameRect.setRight(event->pos().x());

            rectItem->setSelectorFrameBounds(boundingFrameRect);

        }

    }
        break;

    case BottomRight:
    {
        ResizableRectItem * rectItem = dynamic_cast<ResizableRectItem *>( parentItem());
        if(rectItem){

            QRectF boundingFrameRect = rectItem->selectorFrameBounds();

            boundingFrameRect.setRight(event->pos().x());
            boundingFrameRect.setBottom(event->pos().y());

            rectItem->setSelectorFrameBounds(boundingFrameRect);

        }

    }
        break;

    case BottomLeft:
    {
        ResizableRectItem * rectItem = dynamic_cast<ResizableRectItem *>( parentItem());
        if(rectItem){

            QRectF boundingFrameRect = rectItem->selectorFrameBounds();

            boundingFrameRect.setBottom(event->pos().y());
            boundingFrameRect.setLeft(event->pos().x());

            rectItem->setSelectorFrameBounds(boundingFrameRect);

        }

    }
        break;

    }

}
